File: VillagesTown/Data/Army.swift
//
//  Army.swift
//  VillagesTown
//
//  Groups of units that travel together between villages
//

import Foundation

struct Army: Identifiable {
    let id = UUID()
    var name: String
    var units: [Unit]
    var owner: String

    // Current location (village ID or nil if traveling)
    var stationedAt: UUID?

    // Travel info
    var destination: UUID?
    var turnsUntilArrival: Int = 0
    var origin: UUID?

    // Computed properties
    var isMarching: Bool { destination != nil && turnsUntilArrival > 0 }

    var totalAttack: Int {
        units.reduce(0) { $0 + $1.attack }
    }

    var totalDefense: Int {
        units.reduce(0) { $0 + $1.defense }
    }

    var totalHP: Int {
        units.reduce(0) { $0 + $1.currentHP }
    }

    var strength: Int {
        // Combined measure for display
        totalAttack + totalDefense + (totalHP / 10)
    }

    var unitCount: Int { units.count }

    var primaryUnitType: Unit.UnitType? {
        // Most common unit type in army
        let counts = Dictionary(grouping: units, by: { $0.unitType })
        return counts.max(by: { $0.value.count < $1.value.count })?.key
    }

    var emoji: String {
        primaryUnitType?.emoji ?? "⚔️"
    }

    // MARK: - Methods

    mutating func addUnits(_ newUnits: [Unit]) {
        units.append(contentsOf: newUnits)
    }

    mutating func removeDeadUnits() {
        units.removeAll { !$0.isAlive }
    }

    mutating func marchTo(villageID: UUID, turns: Int, from originID: UUID?) {
        self.origin = originID ?? stationedAt
        self.stationedAt = nil
        self.destination = villageID
        self.turnsUntilArrival = turns
    }

    mutating func advanceMarch() {
        if turnsUntilArrival > 0 {
            turnsUntilArrival -= 1
        }
        if turnsUntilArrival == 0, let dest = destination {
            stationedAt = dest
            destination = nil
            origin = nil
        }
    }

    mutating func station(at villageID: UUID) {
        stationedAt = villageID
        destination = nil
        turnsUntilArrival = 0
        origin = nil
    }

    // Generate a name based on composition
    static func generateName(for units: [Unit], owner: String) -> String {
        let count = units.count
        let prefix = owner == "player" ? "" : "Enemy "

        if count <= 3 {
            return "\(prefix)Squad"
        } else if count <= 10 {
            return "\(prefix)Warband"
        } else if count <= 25 {
            return "\(prefix)Company"
        } else {
            return "\(prefix)Legion"
        }
    }
}

// Extension for calculating travel time between villages
extension Army {
    static func calculateTravelTime(from origin: CGPoint, to destination: CGPoint) -> Int {
        let dx = abs(destination.x - origin.x)
        let dy = abs(destination.y - origin.y)
        let distance = sqrt(dx * dx + dy * dy)

        // 1 turn per 5 tiles of distance, minimum 1 turn
        return max(1, Int(ceil(distance / 5.0)))
    }
}

File: villages_town_flutter/lib/data/models/army.dart
import 'dart:math';
import 'dart:ui';
import 'package:uuid/uuid.dart';
import 'unit.dart';
import 'unit_type.dart';

class Army {
  final String id;
  String name;
  List<Unit> units;
  String owner;
  String? stationedAt; // Village ID
  String? destination;
  int turnsUntilArrival;
  String? origin;

  Army({
    String? id,
    required this.name,
    required this.units,
    required this.owner,
    this.stationedAt,
    this.destination,
    this.turnsUntilArrival = 0,
    this.origin,
  }) : id = id ?? const Uuid().v4();

  bool get isMarching => destination != null && turnsUntilArrival > 0;

  int get totalAttack => units.fold(0, (sum, u) => sum + u.attack);
  int get totalDefense => units.fold(0, (sum, u) => sum + u.defense);
  int get totalHP => units.fold(0, (sum, u) => sum + u.currentHP);
  int get strength => totalAttack + totalDefense + (totalHP ~/ 10);
  int get unitCount => units.length;

  UnitType? get primaryUnitType {
    if (units.isEmpty) return null;
    final counts = <UnitType, int>{};
    for (final u in units) {
      counts[u.unitType] = (counts[u.unitType] ?? 0) + 1;
    }
    return counts.entries.reduce((a, b) => a.value > b.value ? a : b).key;
  }

  String get emoji => primaryUnitType?.emoji ?? '⚔️';

  Army copyWith({
    String? id,
    String? name,
    List<Unit>? units,
    String? owner,
    String? stationedAt,
    String? destination,
    int? turnsUntilArrival,
    String? origin,
  }) {
    return Army(
      id: id ?? this.id,
      name: name ?? this.name,
      units: units ?? List.from(this.units),
      owner: owner ?? this.owner,
      stationedAt: stationedAt ?? this.stationedAt,
      destination: destination ?? this.destination,
      turnsUntilArrival: turnsUntilArrival ?? this.turnsUntilArrival,
      origin: origin ?? this.origin,
    );
  }

  void addUnits(List<Unit> newUnits) {
    units.addAll(newUnits);
  }

  void removeDeadUnits() {
    units.removeWhere((u) => !u.isAlive);
  }

  void marchTo(String villageId, int turns, String? fromId) {
    origin = fromId ?? stationedAt;
    stationedAt = null;
    destination = villageId;
    turnsUntilArrival = turns;
  }

  void advanceMarch() {
    if (turnsUntilArrival > 0) {
      turnsUntilArrival--;
    }
    if (turnsUntilArrival == 0 && destination != null) {
      stationedAt = destination;
      destination = null;
      origin = null;
    }
  }

  void station(String villageId) {
    stationedAt = villageId;
    destination = null;
    turnsUntilArrival = 0;
    origin = null;
  }

  static int calculateTravelTime(Offset from, Offset to) {
    final dx = (to.dx - from.dx).abs();
    final dy = (to.dy - from.dy).abs();
    final distance = sqrt(dx * dx + dy * dy);
    return max(1, (distance / 8.0).ceil());
  }

  static String generateName(List<Unit> units, String owner) {
    final prefix = owner == 'player' ? '' : 'Enemy ';
    final count = units.length;
    if (count <= 3) return '${prefix}Squad';
    if (count <= 10) return '${prefix}Warband';
    if (count <= 25) return '${prefix}Company';
    return '${prefix}Legion';
  }
}
